rootchain/sampool/pool.go

```diff
@@ -0,0 +1,146 @@
+package sampool
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/0xPolygon/polygon-edge/rootchain"
+)
+
+var (
+	ErrInvalidHash      = errors.New("invalid SAM hash")
+	ErrInvalidSignature = errors.New("invalid SAM signature")
+	ErrStaleMessage     = errors.New("stale SAM received")
+)
+
+type Verifier interface {
+	VerifyHash(rootchain.SAM) error
+	VerifySignature(rootchain.SAM) error
+
+	Quorum(uint64) bool
+}
+
+type SAMPool struct {
+	verifier Verifier
+
+	mux      sync.Mutex
+	messages map[uint64]samBucket
+
+	lastProcessedIndex uint64
+}
+
+func New(verifier Verifier) *SAMPool {
+	return &SAMPool{
+		verifier: verifier,
+		mux:      sync.Mutex{},
+		messages: make(map[uint64]samBucket),
+	}
+}
+
+func (p *SAMPool) AddMessage(msg rootchain.SAM) error {
+	if err := p.verifySAM(msg); err != nil {
+		return err
+	}
+
+	p.addSAM(msg)
+
+	return nil
+}
+
+func (p *SAMPool) Prune(index uint64) {
+	p.mux.Lock()
+	defer p.mux.Unlock()
+
+	for idx := range p.messages {
+		if idx <= index {
+			delete(p.messages, idx)
+		}
+	}
+
+	p.lastProcessedIndex = index
+}
+
+// TODO: Peek or Pop might be redundant
+func (p *SAMPool) Peek() rootchain.VerifiedSAM {
+	p.mux.Lock()
+	defer p.mux.Unlock()
+
+	expectedIndex := p.lastProcessedIndex + 1
```

Comment:
Is this `Index` sort of `nonce` to be set in original event?

Reply:
Exactly - the `lastProcessedIndex` helps deciding the next event to process from the SAM Pool. (Because the implementation is a `map` and not an ordered queue)